# Estimating and Visualizing Population Genetic Structure for Landscape Genomics
This is a modified version of the example workflow provided by [Caye et al. 2017](https://bcm-uga.github.io/TESS3_encho_sen/articles/main-vignette.html). This is presented as part of UC Berkeley's Center for Computational Biology's Core Skills Seminar Series, a graduate student hosted workshop series focused on dissemintating knowledge of bioinformatics to fellow peers.

TESS3 ([Caye et al. 2016](https://doi.org/10.1111/1755-0998.12471)) is a program for estimating population structure using genome-wide data. TESS3 is similar to STRUCTURE ([Pritchard et al. 2000](10.1093/genetics/155.2.945)) in that it estimates ancestry coefficients based on an optimal number of K populations. However, unlike STRUCTURE, which estimates ancestry coefficients using stochastic Markov chain Monte Carlo methods, or ADMIXTURE, which provides estimates based on maximum likelihood methods, TESS3 generates model-free geographically constrained least-squares estimates of ancestry coefficents.

TESS3 is available as R package [`tess3r`](https://github.com/bcm-uga/TESS3_encho_sen) primarily uses the following types of input data:  
* Genomic data
* Genotype matrix
* Geographic data (e.g. longitude and latitude)

In this example dataset, we will be using data for ~25,000 SNPs from 170 *Arabidopsis thaliana* plants.

## Load Packages and Explore Data
### Load packages:
```{r, echo=FALSE}
rm(list=ls())
library(tess3r)
library(maps)
library(rworldmap)
library(Rgraphviz)
```

### Assign variables for convenience:
```{r, echo=FALSE}
data(data.at)
genotype = data.at$X
coordinates = data.at$coord
```

### Plot coordinates on a map:
```{r, echo=FALSE}
plot(coordinates, pch = 19, cex = .5, 
     xlab = "Longitude (°E)", ylab = "Latitude (°N)")
map(add = T, interior = F)
```

### Questions:
* *What is the spatial extent (scale) of this data?*  
* *What do rows and columns represent?*  
* *How many SNPs are we analyzing?*  

## Running TESS3 
### TESS3 function and arguments:
`tess3r` will take ~10-15 minutes to run on most laptops
* `X`: Genotype matrix
* `coord`: Coordinate data
* `K`: Number of possible ancestral populations we’d like to compare
* `method`: Least squares algorithm used (default as described is “projected.ls”)
* `ploidy`: Ploidy of organism

```{r, echo=FALSE}
tess3.obj <- tess3(X = genotype, 
                   coord = coordinates, 
                   K = 1:8, 
                   method = "projected.ls", 
                   ploidy = 1, 
                   openMP.core.num = 4)
```
### Cross-validation plots:
The optimal number of ancestral populations (K) is determined using cross-entropy cross-validation criterion (Frichot et al. 2014). TESS3 partitions the genotypic matrix into training and test sets and compares predicted genotypic frequencies between observed sets at each locus. Smaller values imply better estimates in TESS3.

```{r, echo=FALSE}
plot(tess3.obj, pch = 19, col = "blue",
     xlab = "Number of ancestral populations",
     ylab = "Cross-validation score")
```

Retrieving data based on desired number of ancestral populations (e.g., tess3 Q matrix for K = 5 clusters)
```{r, echo=FALSE}
q.matrix <- qmatrix(tess3.obj, K = 5)
```

### Questions:
* *What is the purpose of retrieving the 'q.matrix'? What does this data contain?*
* *Do you agree with the chosen number of ancestral populatons?*

## STRUCTURE-like plots:
STRUCTURE, ADMIXTURE, and similar approaches often represent ancestry coefficients with box plots, where colors represent ancestral populations and columns represent individuals. If geographic data is available, box plots may be converted to pie charts layered on a map. TESS3 output can similarly be represented this way.

### Box plots: 
```{r, echo=FALSE}
my.colors <- c("red","yellow","green","blue","violet")
my.palette <- CreatePalette(my.colors, 9)
barplot(q.matrix, border = NA, space = 0, 
        main = "Ancestry matrix", 
        xlab = "Individuals", ylab = "Ancestry proportions", 
        col.palette = my.palette) -> bp
axis(1, at = 1:nrow(q.matrix), labels = bp$order, las = 3, cex.axis = .4)
```

### Pie charts:
```{r, echo=FALSE}
plot(coordinates, pch = 19, cex = .5, 
     xlab = "Longitude (°E)", ylab = "Latitude (°N)")
map(add = T, interior = F)

# Loop through each row of q.matrix and coordinates
for (i in 1:nrow(q.matrix)) {
  # Call the pieGlyph function for each row with respective coordinates
  pieGlyph(q.matrix[i, ], coordinates[i, 1], coordinates[i, 2], 
           edges = 200, radius = 0.5, density = NULL, angle = 45, 
           col = my.colors, border = NULL, lty = NULL, 
           main = paste("Ancestry Coefficients - Point", i))
}
```
## Interpolating Ancestry Coefficients Using `plot`
Q-matrix values generated by `tess3r` can be interpolated on a geographic map using the `plot` function in `R`.

```{r, echo=FALSE}
plot(q.matrix, coordinates, method = "map.max", interpol = FieldsKrigModel(10),  
     main = "Ancestry coefficients",
     xlab = "Longitude", ylab = "Latitude", 
     resolution = c(300,300), cex = .4, 
     col.palette = my.palette)
```

## Bonus: Genome Scans for Selection
TESS3 can use ancestry coefficients to calculate Fst values, and then generate p-values to identify outliers. The Benjamini-Hochberg algorithm ([Bejamini & Hochberg 1995](https://doi.org/10.1111/j.2517-6161.1995.tb02031.x)) is used to control for false-discovery rate.

### Retrieve tess3 results for K = 5:
```{r, echo=FALSE}
p.values <- pvalue(tess3.obj, K = 5)
hist(p.values, col = "lightblue")
```

### Identify candidates by adjusting for false-discovery rate (uses Benjamin-Hochberg algorithm):
```{r, echo=FALSE}
L = length(p.values)
fdr.level = 1e-4
w = which(sort(p.values) < fdr.level * (1:L)/L)
candidates = order(p.values)[w]
length(candidates)
```

### Manhattan plot (outliers highlighted in blue):
```{r, echo=FALSE}
plot(p.values, main = "Manhattan plot", 
     xlab = "Locus id", 
     ylab = "-log10(P-values)",
     cex = .3, col = "grey")
points(candidates, -log10(p.values)[candidates], 
       pch = 19, cex = .2, col = "blue")
```
